### DISCUSSION TITLE
In the case of use `union` or `or` with an object it only parse to the first schema

### DISCUSSION THREAD

**Discussion by leosuncin** [colinhacks/zod#1801] [Q&A] [answered]
I have the following code:

```typescript
import { z } from 'zod';

export const todoSchema = z.object({
  id: z.string().uuid(),
  title: z.string().trim().min(1),
  completed: z.boolean().default(false),
  order: z.number().positive().default(1),
});

export const updateTodoSchema = z.union([
  todoSchema.pick({ completed: true }),
  todoSchema.pick({ order: true }),
  todoSchema.pick({ title: true }),
]);
```

I expect to receive an object with only one property to update `{ completed: boolean }` or `{ order: number }` or `{ title: string }`

But only the object with `completed` gets parsed, in the other cases the property is deleted and set `completed` to its default value

```typescript
updateTodoSchema.parse({ order: 2 });
// { completed: false }
```

## How to reproduce

Open https://stackblitz.com/edit/zod-union?file=src/schemas.spec.ts

In the console run:
```
npm run build; npm test
```

**Comment by JacobWeisenburger** [comment, marked as answer]
How `z.union` works is it loops over each schema option and tries to parse, and it returns the first schema that parses successfully.

Since `completed: z.boolean().default( false )` has a default value, that means `completed` is also optional, and since `updateTodoSchema.parse( { order: 2 } )` doesn't provide a value for `completed`, it will use the default value, which is `false`. So then because the schema `todoSchema.pick({ completed: true })` parsed successfully, it is used as the schema for the union and the others aren't run. And by default `z.object` removes unknown keys, so the `order` key is removed and you are returned `{ completed: false }`.

I think this should give you what you were looking for:

```ts
const todoSchema = z.object( {
    id: z.string().uuid(),
    title: z.string().trim().min( 1 ),
    completed: z.boolean().default( false ),
    order: z.number().positive().default( 1 ),
} )

const updateTodoSchema = todoSchema.pick( {
    completed: true,
    order: true,
    title: true
} ).partial()

const result = updateTodoSchema.safeParse( { order: 2 } )
result.success
    ? console.log( result.data )
    : console.log( result.error.format() )
```
